src/main/java/com/datatheorem/mobileappsecurity/jenkins/plugin/SendBuildAction.java

```diff
@@ -240,27 +243,60 @@ SendBuildMessage uploadBuild(String buildPath) {
                     "Data Theorem upload build returned an error: IOException: " + e.getMessage()
             );
 
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+            return new SendBuildMessage(
+                    false,
+                    "Data Theorem upload build returned an error: IOException: " + e.getMessage()
+            );
         }
     }
 
-    HttpResponse uploadBuildRequest(String buildPath) throws IOException {
+    HttpResponse uploadBuildRequest(String buildPath, Boolean isArtifact) throws IOException, InterruptedException {
         /*
          * Http call of the upload link generated by upload_init
          * @return:
          *   The HTTPResponse of the endpoint
          */
 
-        HttpClient client = HttpClientBuilder.create().build();
-        HttpEntity entity = MultipartEntityBuilder.create().addBinaryBody("file", new File(buildPath)).build();
 
         // Create an http client to make the post request to upload_init endpoint
+
+        HttpClient client = HttpClientBuilder.create().build();
         HttpPost requestUploadbuild = new HttpPost(this.uploadUrl);
+
+        if (isArtifact) {
+            // if the build is in the permanent artifact directory we can upload it directly
+            HttpEntity entity = MultipartEntityBuilder.create().addBinaryBody("file", new File(buildPath)).build();
+            requestUploadbuild.setEntity(entity);
+
+            // Add the api access key of the customer and tell to Upload API that the request comes from jenkins
+            requestUploadbuild.addHeader("User-Agent", "Jenkins Upload API Plugin " + version);
+            HttpResponse response = client.execute(requestUploadbuild);
+            logger.println(response.getStatusLine().toString());
+            return response;
+        }
+
+        // if the build is in the workspace we copy the content in an Output Stream and we upload the stream
+
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        workspace.child(buildPath).copyTo(os);
+        logger.println(os.toByteArray().length);
+
+        // send the file
+        HttpEntity entity = MultipartEntityBuilder.create().addBinaryBody(
+                "file",
+                os.toByteArray(),
+                ContentType.DEFAULT_BINARY,
+                workspace.child(buildPath).getName()
+        ).build();
         requestUploadbuild.setEntity(entity);
 
         // Add the api access key of the customer and tell to Upload API that the request comes from jenkins
         requestUploadbuild.addHeader("User-Agent", "Jenkins Upload API Plugin " + version);
         HttpResponse response = client.execute(requestUploadbuild);
-        logger.println(response.getStatusLine().toString());
+        os.flush();
+        os.close();
```

Comment:
Can there be an exception thrown before this cleanup step that will cause the os to not be flushed/closed?

Reply:
Yes i guess it can raise an IO exception